plat/kvm/x86/lcpu_start.S

```diff
@@ -0,0 +1,241 @@
+	orl 	$X86_EFER_LME, %eax
+	orl 	$X86_EFER_NXE, %eax
+	wrmsr
+
+	/* 3: Load PML pointer */
+	movl 	cr3_val, %eax
+	movl 	%eax, %cr3
+
+	/* 4: Enable paging */
+	movl	%cr0, %eax
+	orl	$X86_CR0_PG, %eax
+	movl	%eax, %cr0
+
+	lgdt 	RELOC(gdt64_ptr)
+	ljmp 	$(gdt64_cs-gdt64), $RELOC(_lcpu_start64)
+END(_lcpu_start32)
+.align 64
+
+.code64
+.align 32
+ENTRY(_lcpu_start64)
+	/* Wait for the BSP to finish */
+spin:
+	pause
+	cmpb 	$0, bspdone
+	jz 	spin
+
+	/* get the AP id */
+	mov	$1, %eax
+	cpuid
+	shrl	$24, %ebx
+
+	/* Notify the BSP that the core has started successfully */
+	lock	incq smp_aprunning
+
+	/* Taken from entry64.S */
+	movq 	%rbx, %r9
+	movq 	%cr0, %rsi
+	movq 	%cr4, %rdi
+
+	/* FPU and SSE are part of base x86-64, so no need to check for their
+	 * availability before enabling and initializing. */
+	andl 	$(~(X86_CR0_EM | X86_CR0_TS)), %esi
+	orl 	$(X86_CR0_MP | X86_CR0_NE | X86_CR0_WP), %esi
+	movq 	%rsi, %cr0
+	fninit
+#if __SSE__
+	orl 	$(X86_CR4_OSFXSR | X86_CR4_OSXMMEXCPT), %edi
+	movq 	%rdi, %cr4
+	ldmxcsr (mxcsr_ptr)
+#endif /* __SSE__ */
+#if (__AVX__ || CONFIG_HAVE_X86PKU)
+	/* Check capabilities subject to availability as indicated by cpuid.
+	 * First, start off with "standard features" */
+	movl 	$0x1, %eax
+	cpuid
+	/* ecx and edx now contain capability information, so we can now
+	 * enable capabilities based on the indicated features */
+	/* note: OSXSAVE needs to be enabled before AVX and PKU */
+	testl 	$(X86_CPUID1_ECX_XSAVE), %ecx
+	jz 	noxsave
+	orl 	$(X86_CR4_OSXSAVE), %edi
+	movq 	%rdi, %cr4
+#if __AVX__
+	/* now enable AVX. This needs to be last checking cpuid features from
+	 * the eax=1 cpuid call, because it clobbers ecx */
+	testl 	$(X86_CPUID1_ECX_AVX), %ecx
+	jz 	noavx
+	xorl 	%ecx, %ecx
+	xgetbv
+	orl 	$(X86_XCR0_SSE | X86_XCR0_AVX), %eax
+	xsetbv
+noavx:
+#endif /* __AVX__ */
+/* Do not enable AVX without XSAVE, otherwise we'll get #UD */
+noxsave:
+#endif /* __AVX__ || CONFIG_HAVE_X86PKU */
+	/* Now, check for extended features. */
+	movl 	$0x7, %eax
+	movl 	$0x0, %ecx
+	cpuid
+	/* ebx, ecx, edx now contain extended capabilities information. */
+	/* check for and enable FS/GSBASE */
+	testl 	$(X86_CPUID7_EBX_FSGSBASE), %ebx
+	jz 	nofsgsbase
+	orl $	(X86_CR4_FSGSBASE), %edi
+	movq 	%rdi, %cr4
+nofsgsbase:
+#if CONFIG_HAVE_X86PKU
+	/* check for Memory Protection Keys (PKU) */
+	testl 	$(X86_CPUID7_ECX_PKU), %ecx
+	jz 	nopku
+	/* only enable PKU if we support XSAVE */
+	testl 	$(X86_CR4_OSXSAVE), %edi
+	jz 	nopku
+	/* PKU is supported, enable it via CR4 */
+	orl 	$(X86_CR4_PKE), %edi
+	movq 	%rdi, %cr4
+	/* also enable XSAVE for the PKRU */
+	xorl 	%ecx, %ecx
+	xgetbv
+	orl 	$(X86_XCR0_PKRU), %eax
+	xsetbv
+nopku:
+#endif /* CONFIG_HAVE_X86PKU */
+	/* Load IDT */
+	lidt 	(idtptr)
+
+	/* Use CPU ID * cpustruct_size for indexing the cpu structure */
+	movq 	%r9, %rax
+	movq	cpustruct_size, %r10
+	mulq 	%r10
+	movq 	%rax, %r10
+
+	/* Load the stack pointer and the entry address */
+	movq	cpusptr, %rsi

```

Comment:
use the `lea` instruction and the actual cpu struct array symbol to calculate the address. Delete the `cpusptr` variable.